cpp/src/arrow/array/array_list_test.cc

```diff
@@ -509,6 +509,18 @@ class TestListArray : public ::testing::Test {
     ASSERT_RAISES(Invalid, ValidateOffsets(2, {0, 7, 4}, values));
   }
 
+  void TestSliced() {
+    auto arr = ArrayFromJSON(list(int16()), "[[1, 2], [3, 4, 5], [6], [7, 8]]");
+
+    auto arr_sliced = arr->Slice(0, 2);
+    auto expected_sliced = ArrayFromJSON(list(int16()), "[[1, 2], [3, 4, 5]]");
+    AssertArraysEqual(*expected_sliced, *arr_sliced);
+
+    auto values = checked_cast<ListArray*>(arr_sliced.get())->values();
+    auto expected_values = ArrayFromJSON(int16(), "[1, 2, 3, 4, 5]");
+    AssertArraysEqual(*expected_values, *values);
```

Comment:
Should we make this expected behavior or not?

Reply:
I just commented on the JIRA as well, but so in the current design, we shouldn't change that. It is `Flatten()` that already implements this, and the `values` being unsliced ensures that they still match the `offsets` of a sliced ListArray.

It is certainly confusing, though. I wonder if we should give it a more scary name like "raw_values"

Reply:
> The problem with Flatten though is that it removes the null values from the values array, but in this case we want them.

Because we use the offsets to slice into the flat values, and those offsets take into account potential values behind a null?

This corner case is covered by existing tests?

Reply:
I should add a test.

And you might be right about flatten, it does include the nulls in the values (not the list). For some reason earlier I had gotten the impression it doesn't, while I was debugging.

```python
>>> import pyarrow as pa
>>> arr = pa.array([[1, 2], [3, 4, 5], [6, None], [7, 8]])
>>> arr.flatten()
<pyarrow.lib.Int64Array object at 0x129b2c820>
[
  1,
  2,
  3,
  4,
  5,
  6,
  null,
  7,
  8
]
```

Reply:
Yes, but your example uses an actual null inside a list, not a "null list". For example, here the first null is not in the flattened output:

```
In [3]: arr = pa.array([[1, 2], None, [3, None]])

In [4]: arr.flatten()
Out[4]: 
<pyarrow.lib.Int64Array object at 0x7f8b1fbcaa40>
[
  1,
  2,
  3,
  null
]
```

And I suppose behind that null can be any data (the default when constructed above is that there is no data behind, so the offset doesn't increment for that list element):

```
In [5]: arr.offsets
Out[5]: 
<pyarrow.lib.Int32Array object at 0x7f8b4f0f5a80>
[
  0,
  2,
  2,
  4
]
```

It's a bit tricky to construct manually, but something like:

```
In [10]: arr = pa.ListArray.from_arrays(pa.array([0, 2, 4, 6]), pa.array([1, 2, 99, 99, 3, None]), mask=pa.array([False, True, False]))

In [11]: arr
Out[11]: 
<pyarrow.lib.ListArray object at 0x7f8b1fbcba60>
[
  [
    1,
    2
  ],
  null,
  [
    3,
    null
  ]
]

In [12]: arr.flatten()
Out[12]: 
<pyarrow.lib.Int64Array object at 0x7f8b4f065960>
[
  1,
  2,
  3,
  null
]

In [13]: arr.values
Out[13]: 
<pyarrow.lib.Int64Array object at 0x7f8b4f065780>
[
  1,
  2,
  99,
  99,
  3,
  null
]

In [14]: arr.offsets
Out[14]: 
<pyarrow.lib.Int32Array object at 0x7f8b1fbc9300>
[
  0,
  2,
  4,
  6
]
```

But, so I am not sure if for this case you actually need to flattened (with nulls removed) or not for this case of converting to numpy.

Reply:
> Because we use the offsets to slice into the flat values, and those offsets take into account potential values behind a null?

Oh now I remember, and I think I understand what you are saying here better now. For the fixed size lists, the values behind a null entry in the list are removed when we call `Flatten()`. When we go back to try to reconstruct the lists based on offsets, the offsets produced by `value_offset` are all invalid since they don't account for the values we dropped in `Flatten()`.